core/src/main/scala/io/odin/formatter/Formatter.scala

```diff
@@ -127,4 +154,18 @@ object Formatter {
         writeStackTrace(b, elements.tail, indent)
     }
   }
+
+  private def abbreviate(enclosure: String): String = {
+    @tailrec
+    def loop(input: List[String], builder: StringBuilder): StringBuilder = {
+      input match {
+        case Nil          => builder
+        case head :: Nil  => builder.append(head)
+        case head :: tail => loop(tail, builder.append(head.headOption.getOrElse('?')).append('.'))
+      }
+    }
+
+    loop(enclosure.split('.').toList, new StringBuilder).toString()
```

Comment:
You can avoid bouncing over `Char` by using string right away, as under the hood it still allocates yet another string object.

Also, I'd stay with `Array` and just check its size in `loop` to avoid `List` conversion

Reply:
Also, it would make sense to move `"."` to some constant `private val dot` inside of `Formatter` object to avoid spawning those `String` instances on each call.